text/0000-stack-buildpacks.md

```diff
@@ -0,0 +1,596 @@
+
+ ```bash
+#!/usr/bin/env bash
+
+exit 0
+```
+
+Its `bin/build` would have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+apt update -y
+
+for package in $(cat $3 | yj -t | jq -r ".entries | .[] | .name"); do
+  apt install $package
+done
+
+cat << EOF > $1/launch.toml
+[[excludes]]
+paths = [ "/var/cache" ]
+cache = true
+EOF
+```
+
+## Example: CA Certificate Buildpack
+
+Support for custom CA Certificates can be accomplished with two buildpacks: a stackpack that can install the cert, and a normal buildpack that can provide a cert in the build plan.
+
+### CA Cert Installer Stackpack
+
+A buildpack that installs custom CA Certificates would have a `buildpack.toml` that looks like this:
+
+```toml
+[buildpack]
+id = "example/cacerts"
+privileged = true
+
+[[stacks]]
+id = "io.buildpacks.stacks.bionic"
+```
+
+Its `bin/detect` would have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+cat << EOF > $2
+[[provides]]
+name = "cacert"
+EOF
+
+exit 0
+```
+
+Its `bin/build` would would install each `cacert` in the build plan. It would have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+# filter to the cert
+for file in $(cat $3 | yj -t | jq -r ".entries | .[] | select(.name==\"cacert\") | .metadata | .file"); do
+  cert="$(cat $3 | yj -t | jq -r ".entries | .[] | select(.name==\"cacert\") | .metadata | select(.file==\"$file\") | .content")"
+  echo $cert > /usr/share/ca-certificates/$file
+  chmod 644 /usr/share/ca-certificates/$file
+done
+
+update-ca-certificates
+```
+
+### CA Cert Provider Buildpack
+
+The stackpack must be used with a buildpack that provides a certificate(s) for it to install. That buildpack would have the following `buildpack.toml`:
+
+```toml
+[buildpack]
+id = "my/db-cert"
+
+[[stacks]]
+id = "io.buildpacks.stacks.bionic"
+```
+
+Its `bin/detect` would require a certificate with the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+cat << EOF > $2
+[[requires]]
+name = "cacert"
+
+[requires.metadata]
+file = "database.crt"
+content = """
+$(cat $CNB_BUILDPACK_DIR/database.crt)
+"""
+
+[[requires]]
+name = "cacert"
+
+[requires.metadata]
+file = "server.crt"
+content = """
+$(cat $CNB_BUILDPACK_DIR/server.crt)
+"""
+EOF
+
+exit 0
+```
+
+Its `bin/build` would do nothing, and `have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+exit 0
+```
+
+## Example: jq mixin
+
+A buildpack may that requires the `jq` package may have it provided by either a stack, stack buildpack, or userspace buildpack.
+
+```toml
+[[or.requires]]
+name = "jq"
+mixin = true
+
+[[or.requires]]
+name = "jq"
+```
+
+## Future work
+
+In the future, we plan to enhance the stack buildpack interface with the following:
+
+* A `CNB_STACK_TYPE` env var that a stack buildpack can use to behave differently on each part of the stack
+* Support for `creator`. Because of the new extend phase, it is not possible to easily run the entire buildpack process with a single binary.
```

Comment:
Maybe it's worth calling out that `creator` can still run with stack buildpacks if there is no need to extend the run image. I believe this is true looking at https://github.com/buildpacks/lifecycle/pull/410 , but maybe I'm wrong.

Reply:
@jabrown85 is it true that you can use `creator` if you don't extend the run image, or will `creator` just ignore stackpacks?

Reply:
You can use `creator` as long as it doesn't detect you need to extend the run image.

Reply:
But we don't have to keep that functionality if it makes things more confusing. I did that work prior to realizing `creator` wasn't going to work for extend images.